rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+                ptr::null_mut(),
+                0,
+                num_format.to_num_format(),
+                key_usage.to_flag(),
+                rsa_key,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaPublicKey {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                    number_format: num_format,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `export_public_key_blob()` will export a [`RsaPublicKey`].
+    pub fn export_public_key_blob(&self) -> Result<RsaPublicKeyBlob, SymCryptError> {
+        let mut modulus_buffer = vec![0u8; self.size_of_modulus() as usize];
+        let mut pub_exp = vec![0u64; 1];
+        unsafe {
+            // SAFETY: FFI calls
+            // When only getting the public key, ppPrimes, pcbPrimes and nPrimes can be NULL, NULL and 0.
+            match symcrypt_sys::SymCryptRsakeyGetValue(
+                self.inner(),
+                modulus_buffer.as_mut_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                pub_exp.as_mut_ptr(),
+                1 as symcrypt_sys::UINT32,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0,
+                self.number_format.to_num_format(),
+                self.key_usage.to_flag(),
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaPublicKeyBlob {
+                    modulus: modulus_buffer,
+                    pub_exp: self.store_msb_first_u64(pub_exp[0])?,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `number_of_pub_primes()` returns a `u32` representing the number of primes associated with the [`RsaPublicKey`]
+    pub fn number_of_pub_primes(&self) -> u32 {
+        0
+        // !REVIEW: Can call SymCryptRsakeyGetNumberOfPrimes() instead, but it's a wasteful call since we know RsaPublicKey must have value 0.
+        // This value is assumed since we know that this object is a RsaPublicKey
+    }
+}
+
+impl Drop for RsaPublicKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeyFree(self.inner());
+        }
+    }
+}
+
+// Utility function to reduce common RSA allocation call
+fn allocate_rsa(n_primes: u32, n_bits_mod: u32) -> symcrypt_sys::PSYMCRYPT_RSAKEY {
+    let rsa_params = symcrypt_sys::SYMCRYPT_RSA_PARAMS {
+        version: 1 as symcrypt_sys::UINT32, // No other version aside from version 1 is specified
+        nBitsOfModulus: n_bits_mod as symcrypt_sys::UINT32,
+        nPrimes: n_primes as symcrypt_sys::UINT32, // To generate a only a public key, 0 must be passed.
+        nPubExp: 1 as symcrypt_sys::UINT32,        // Value must be 1.
+    };
+    unsafe {
+        // SAFETY: FFI calls
+        symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0)
```

Comment:
>  symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0) [](http://example.com/codeflow?start=6&length=54)

This function can fail and return NULL.

Reply:
`PSYMCRYPT_RSAKEY
SYMCRYPT_CALL
SymCryptRsakeyAllocate(
    _In_    PCSYMCRYPT_RSA_PARAMS   pParams,
    _In_    UINT32                  flags );
//
// Allocate and create a new RSAKEY object sized according to the parameters.
// If the SYMCRYPT_RSAKEY object will only be used for a public key, the
// SYMCRYPT_RSA_PARAMS structure may set nPrimes = 0.
//
// This call does not initialize the key. It should be
// followed by a call to SymCryptRsakeyGenerate or
// SymCryptRsakeySetValue.
//
// No flags are specified for this function.
//`

No indication that it can return NULL or fail

Reply:
NTSTATUS
AllocateSymcryptRsakey(
    __inout MSCRYPT_RSA_KEY *pKey)
{
    NTSTATUS            Status = STATUS_SUCCESS;
    SYMCRYPT_RSA_PARAMS SymcryptRsaParam;

    SymcryptRsaParam.version = 1;
    SymcryptRsaParam.nBitsOfModulus = pKey->dwModBitLen;
    SymcryptRsaParam.nPrimes = 2;
    SymcryptRsaParam.nPubExp = 1;

    pKey->Key = SymCryptRsakeyAllocate(&SymcryptRsaParam, 0);
    if (pKey->Key == NULL)
    {
        Status = STATUS_NO_MEMORY;
        TRACE_ERROR_STATUS(Status);
        goto cleanup;
    }

cleanup:

    return Status;
}

Reply:
Fixed